packages/SwingSet/docs/virtual-objects.md

```diff
@@ -0,0 +1,111 @@
+
+```javascript
+const fooCounter = counterMaker('foo');
+const barCounter = counterMaker('bar');
+
+fooCounter.inc();
+fooCounter.inc();
+barCounter.inc();
+barCounter.rename('new bar');
+console.log(`${fooCounter.getName()} count is ${fooCounter.getCount()`); // "foo count is 2"
+console.log(`${barCounter.getName()} count is ${barCounter.getCount()`); // "new bar count is 1"
+```
+
+Additional important details:
+
+- The set of state properties is determined by the `initialize` method.  The properties that exist when `initialize` returns _are_ the properties of the object.  State properties cannot thereafter be added or removed.
+
+- The values a state property may take are limited to things that are serializable and which may be hardened and made immutable (and, in fact, _are_ hardened and made immutable the moment they are assigned).  That is, you can replace what value a state property _has_, but you cannot modify a state property's value in place.  In other words, you can do things like:
+
+  ```state.zot = [1, 2, 3];```
+
+  but you can't do things like:
+
+  ```state.zot.push(4);```
+
+- A virtual object can be passed as a parameter in messages to other vats.  It will be passed by presence, just like any other non-data object you might send in a message parameter.
+
+- A virtual object's state may include references to other virtual objects.
+
+- The in-memory manifestations of a virtual object (which we call its "representatives") are not necessarily (or even usually) `===` to each other, even when they represent the same object, since a new representative is created for each act of deserialization.  They do, however, all operate on the same state, so changes to the state made by a method call to one representative will be visible to the other representatives.  A soon-to-arrive future version of the virtual object system will add a `sameKey` predicate that will allow you to compare two virtual object representatives to determine if they refer to the same virtual object instance.
+
+## Persistent Weak Stores
+
+The vat secondary storage system also includes a `makeWeakStore` function, which is identical in API to the `makeWeakStore` function provided by the `@agoric/store` package.  I.e., you call:
+
+`const ws = makeWeakStore();`
+
+to obtain a new weak store instance, which you can then access using the `has`, `init`, `set`, `get`, and `delete` methods just as you would if you had imported `makeWeakStore` from `@agoric/store`.
+
+However, the vat secondary storage system's `makeWeakStore` augments the one provided by the `@agoric/store` package in two important ways:
+
+- Weak store operations may use virtual object representatives as keys, and if you do this the corresponding underlying virtual object instance identities will be used for indexing rather than the object identities of the representative objects.  (I.e., two object representives for which `sameKey` is true will operate on the same weak store entry).
```

Comment:
I thought we decided to fix this using WeakRefs?

Reply:
Forgive me, but I totally don't understand what you mean by that.  Fix what using weakrefs how?

Reply:
Ah, I recall the discussion now.  That would allow us to reduce the current 3-layer implementation to a 2-layer implementation, which would simplify some things (though it would also require doing delicate `WeakRef` things in exchange, so I'm not sure it's a net win on implementation complexity -- though enabling users to use `===` probably pays for it).  Did we eventually decide we were OK committing to Node 14 as the minimum required version?

BTW, I've implemented a placeholder `sameKey` function in the virtual object manager (PR forthcoming) but it only understands the inner key magic for virtual objects (everything else goes through object identity) and I was wondering about the work of integrating it with the current `sameStructure` implementation.  Presumably this effort can disappear in a puff of logic if `===` works?

Reply:
It would indeed disappear.

It's not just the user using `===`. It is all the hidden ways users depend on JS object identity which we'd need to teach them to avoid.

SES-shim needs to support earlier versions of Node. agoric-sdk does not. Endo will be somewhere between the two attn @kriskowal . This persistence system is right on the edge between Endo and agoric-sdk. Not sure yet which way it should go.

Reply:
This persistence system is very tightly integrated with the SwingSet kernel and liveslots, so I don't think it winds up as an Endo thing per se.  It's plausible we'll want something like it for Endo, and also plausible that the something like it, whatever it proves to be, can cannibalize the current implementation, but I think we first have to answer some questions about the role of CapTP in Endo.  Off topic here, I think.

Reply:
@warner I think we need to have a conversation about the priority of reworking things using weak refs, relative to the other things on our plate.

Reply:
Since we now have the option of mapping one-to-one with JS object identity, I strongly encourage us to.